File: pset1/credit.c
#include <stdio.h>
#include <cs50.h>
#include <stdlib.h>

long asknumber(void);
int checknumber_length(long n);
void checkCard(long n, int num_digits);

int main(void)
{
    
    long number = asknumber();
    int digits = checknumber_length(number);
    checkCard(number, digits);
    
}
//--------------------------------------------------------
long asknumber(void)
{
    printf("Number: ");
    long x = get_long_long();
    
    return x;
}
//--------------------------------------------------------
int checknumber_length(long n)
{
    int num_digits = 0;
    while (n>=1)
    {
        num_digits = num_digits + 1;
        n = n/10;
        
    }
    
    return num_digits;
}
//--------------------------------------------------------
void checkCard(long n, int num_digits)
{
    ///Extract digits into an array
    int numberArray[num_digits];
    int c = 0;
    while (n != 0)
    {
        numberArray[c] = n % 10;
        n /= 10;
        c++;
    }

    //Step  0
    long sum = 0;
    long secondSum = 0;
    
    for (int i=0; i < num_digits; i=i+1)
    {
        if (i%2 == 1)
        {
            if (2*numberArray[i]>9)
            {
                sum = sum + 1;
                sum = sum + 2*(numberArray[i])%10;
            }
            else
            {
                sum = sum + 2*numberArray[i];
            }
        }
        
       else
        {
            secondSum = secondSum + numberArray[i];
        }
    }
    
    if ((sum+secondSum)%10 == 0)
    {
        if (numberArray[num_digits-1] == 3 && (numberArray[num_digits-2] == 4 || numberArray[num_digits-2] == 7) && num_digits == 15)
        {
            printf("AMEX\n");
        }
        
        else if (numberArray[num_digits-1] == 5 && (numberArray[num_digits-2] == 1 || numberArray[num_digits-2] == 2 || numberArray[num_digits-2] == 3 || numberArray[num_digits-2] == 4 || numberArray[num_digits-2] == 5) && num_digits == 16)
        {
            printf("MASTERCARD\n");
        }
        
        else if (numberArray[num_digits-1] == 4 && (num_digits == 13 || num_digits == 16))
        {
            printf("VISA\n");
        }
        
        else
        {
            printf("INVALID\n");
        }
    }
    
    
    else
    {
        printf("INVALID\n");
    }
    
}
//--------------------------------------------------------

File: pset6/credit.py
def main():
    
    creditcard = input("Number: ")
    
    while creditcard.isnumeric() == False:
        creditcard = input("Retry: ")
    
    checkCard(creditcard)
    
def checkCard(number):
    """Takes in credit card number.
    Outputs a print string to terminal saying if it is a card
    If it is a card it will give the type of credit card
    If it is not a card, it will print INVALID
    """
    #convert number into an array for operations
    number_array = []
    for i in number:
        number_array.append(int(i))

    # Multiply every other digit by 2, starting with the number’s second-to-last digit
    l = number_array[::-1]
    l[1::2] = [i*2 for i in l[1::2]]
    
    # Add all the digits together!
    summation = 0
    for i in l:
        # if the number is over 10, we need to add its digits to get the proper sum
        if i>9:
            i = i//10 + i%10
        summation = summation+i
    
    # If the total’s last digit is 0 (or, put more formally, if the total modulo 10 is congruent to 0), the number is valid!
    if summation%10 == 0:
        
        # American Express numbers all start with 34 or 37
        if int(number[0]+number[1]) in {34, 37}:
            print("AMEX")
        
        # MasterCard numbers all start with 51, 52, 53, 54, or 55;
        elif int(number[0]+number[1]) in {51, 52, 53, 54, 55}:
            print("MASTERCARD")
        
        # Visa numbers all start with 4.
        elif number[0] == 4:
            print("VISA")
            
        else:
            print("INVALID")
    
    else:
        print("INVALID")
    
if __name__ == "__main__":
    main()